conbench/entities/benchmark_result.py

```diff
@@ -507,7 +459,122 @@ def ui_hardware_short(self) -> str:
         return f"{hw.id[:4]}: " + hw.name
 
 
-def do_iteration_samples_look_like_error(data: list[Optional[Decimal]]) -> bool:
+def validate_and_aggregate_samples(stats_usergiven: Any):
+    """
+    Raises BenchmarkResultValidationError upon logical inconsistencies.
+
+    Only run this for the 'success' case, i.e. when the input is a list longer
+    than zero, and each item is a number (not math.nan).
+
+    This returns a dictionary with key/value pairs meant for DB insertion,
+    top-level for BenchmarkResult.
+    """
+
+    agg_keys = ("q1", "q3", "mean", "median", "min", "max", "stdev", "iqr")
+
+    # Encode invariants.
+    samples = stats_usergiven["data"]
+    assert len(samples) > 0
+    for sa in samples:
+        assert sa is not None
+        assert sa != math.nan
+
+    # Proceed with float type.
+    samples = [float(s) for s in samples]
+
+    # First copy the entire stats data structure (this includes times, data,
+    # mean, min, ...). Later: selectively overwrite/augment.
+    result_data_for_db = stats_usergiven.copy()
+
+    # Initialize all aggregate values explicitly to None -- values set below
+    # must be meaningful.
+    for k in agg_keys:
+        result_data_for_db[k] = None
+
+    if len(samples) >= 2:
+        # There is a special case where for one sample the iteration count
+        # may be larger than one, see below.
+        if stats_usergiven["iterations"] != len(samples):
+            raise BenchmarkResultValidationError(
+                f'iterations count ({ stats_usergiven["iterations"] }) does '
+                f"not match sample count ({len(samples)})"
+            )
+
+    if len(samples) >= 3:
+        # See https://github.com/conbench/conbench/issues/802 and
+        # https://github.com/conbench/conbench/issues/1118
+        q1, q3 = np.percentile(samples, [25, 75])
+
+        aggregates = {
+            "q1": q1,
+            "q3": q3,
+            "mean": np.mean(samples),
+            "median": np.median(samples),
+            "min": np.min(samples),
+            "max": np.max(samples),
+            # https://numpy.org/doc/stable/reference/generated/numpy.std.html
+            # This has N in the divisor (ddof=0)
+            "stdev": np.std(samples),
+            "iqr": q3 - q1,
+        }
+
+        # Now, overwrite with self-derived aggregates:
+        for key, value in aggregates.items():
+            result_data_for_db[key] = sigfig.round(value, sigfigs=5)
```

Comment:
Is there a reason you chose 5 instead of 4?

Reply:
Good question! I didn't think of a good reason when I did this I think, maybe I did and now I don't remember anymore.